src/app/CommandHandlerImpl.h

```diff
@@ -0,0 +1,488 @@
+
+    CHIP_ERROR FallibleAddStatus(const ConcreteCommandPath & aRequestCommandPath, const Protocols::InteractionModel::Status aStatus,
+                                 const char * context = nullptr) override;
+    void AddStatus(const ConcreteCommandPath & aCommandPath, const Protocols::InteractionModel::Status aStatus,
+                   const char * context = nullptr) override;
+    CHIP_ERROR AddClusterSpecificSuccess(const ConcreteCommandPath & aRequestCommandPath, ClusterStatus aClusterStatus) override;
+    CHIP_ERROR AddClusterSpecificFailure(const ConcreteCommandPath & aRequestCommandPath, ClusterStatus aClusterStatus) override;
+
+    CHIP_ERROR AddResponseData(const ConcreteCommandPath & aRequestCommandPath, CommandId aResponseCommandId,
+                               DataModel::EncodableToTLV & aEncodable) override;
+    void AddResponse(const ConcreteCommandPath & aRequestCommandPath, CommandId aResponseCommandId,
+                     DataModel::EncodableToTLV & aEncodable) override;
+
+    Access::SubjectDescriptor GetSubjectDescriptor() const override;
+    FabricIndex GetAccessingFabricIndex() const override;
+    bool IsTimedInvoke() const override;
+
+    Messaging::ExchangeContext * GetExchangeContext() const override
+    {
+        VerifyOrDie(mpResponder);
+        return mpResponder->GetExchangeContext();
+    }
+
+    /**************** Implementation-specific logic ***********************/
+
+    /*
+     * Main entrypoint for this class to handle an InvokeRequestMessage.
+     *
+     * This function MAY call the registered OnDone callback before returning.
+     * To prevent immediate OnDone invocation, callers can wrap their CommandHandlerImpl instance
+     * within a CommandHandler::Handle.
+     *
+     * isTimedInvoke is true if and only if this is part of a Timed Invoke
+     * transaction (i.e. was preceded by a Timed Request).  If we reach here,
+     * the timer verification has already been done.
+     *
+     * commandResponder handles sending InvokeResponses, added by clusters, to the client. The
+     * command responder object must outlive this CommandHandler object. It is only safe to
+     * release after the caller of OnInvokeCommandRequest receives the OnDone callback.
+     */
+    Protocols::InteractionModel::Status OnInvokeCommandRequest(CommandHandlerExchangeInterface & commandResponder,
+                                                               System::PacketBufferHandle && payload, bool isTimedInvoke);
+
+    /**
+     * Checks that all CommandDataIB within InvokeRequests satisfy the spec's general
+     * constraints for CommandDataIB. Additionally checks that InvokeRequestMessage is
+     * properly formatted.
+     *
+     * This also builds a registry to ensure that all commands can be responded
+     * to with the data required as per spec.
+     */
+    CHIP_ERROR ValidateInvokeRequestMessageAndBuildRegistry(InvokeRequestMessage::Parser & invokeRequestMessage);
```

Comment:
nit, I though in slack we mentioned that this could be moved to private

Reply:
I think some things are not needed public anymore (e.g. all prepare/getwriter/finish). I would prefer to do that as a followup as I am unsure what can or cannot be moved ... some things may be used by unittests.

Reply:
ack